action/protocol/context.go

```diff
@@ -98,7 +98,115 @@ func GetValidateActionsCtx(ctx context.Context) (ValidateActionsCtx, bool) {
 func MustGetValidateActionsCtx(ctx context.Context) ValidateActionsCtx {
 	va, ok := ctx.Value(validateActionsCtxKey{}).(ValidateActionsCtx)
 	if !ok {
-		log.S().Panic("Miss run actions context")
+		log.S().Panic("Miss validate actions context")
 	}
 	return va
 }
+
+// TODO: replace RunActionsCtx and ValidateActionsCtx with below classified independent contexts
+
+type blockchainCtxKey struct{}
+
+type blockCtxKey struct{}
+
+type actionCtxKey struct{}
+
+// BlockchainCtx provides blockchain auxiliary information.
+type BlockchainCtx struct {
+	// Genesis is a copy of current genesis
+	Genesis genesis.Genesis
+	// History indicates whether to save account/contract history or not
+	History bool
+	// Registry is the pointer protocol registry
+	Registry *Registry
+}
+
+// BlockCtx provides block auxiliary information.
+type BlockCtx struct {
+	// height of block containing those actions
+	BlockHeight uint64
+	// timestamp of block containing those actions
+	BlockTimeStamp time.Time
+	// gas Limit for perform those actions
+	GasLimit uint64
+	// Producer is the address of whom composes the block containing this action
+	Producer address.Address
+}
+
+// ActionCtx provides action auxiliary information.
+type ActionCtx struct {
+	// Caller is the address of whom issues this action
+	Caller address.Address
+	// ActionHash is the hash of the action with the sealed envelope
+	ActionHash hash.Hash256
+	// GasPrice is the action gas price
+	GasPrice *big.Int
+	// IntrinsicGas is the action intrinsic gas
+	IntrinsicGas uint64
+	// Nonce is the nonce of the action
+	Nonce uint64
+	// History indicates whether to save account/contract history or not
+}
+
+// WithBlockchainCtx add BlockchainCtx into context.
+func WithBlockchainCtx(ctx context.Context, bc BlockchainCtx) context.Context {
+	return context.WithValue(ctx, blockchainCtxKey{}, bc)
+}
+
+// GetBlockchainCtx gets BlockchainCtx
+func GetBlockchainCtx(ctx context.Context) (BlockchainCtx, bool) {
+	bc, ok := ctx.Value(blockchainCtxKey{}).(BlockchainCtx)
```

Comment:
`return ctx.Value(blockchainCtxKey{}).(BlockchainCtx)`

Reply:
interface type assertion needs to be assigned to return 2 values